Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialTypeBuilder.cpp

```diff
@@ -199,6 +268,269 @@ namespace AZ
+                    "#endif                                                                                            \n"
+                    "                                                                                                  \n"
+                    "#if !MaterialFunction_AdjustSurface_DEFINED && MATERIALPIPELINE_SHADER_HAS_PIXEL_STAGE            \n"
+                    "    void MaterialFunction_AdjustSurface(inout Surface outSurface) {}                              \n"
+                    "    #define MaterialFunction_AdjustSurface_DEFINED 1                                              \n"
+                    "#endif                                                                                            \n"
+                    ;
+
+                AZ::IO::Path shaderName = shaderTemplate.m_shader;
+                shaderName = shaderName.Filename(); // Removes the folder path
+                shaderName = shaderName.ReplaceExtension(""); // This will remove the ".template" extension
+                shaderName = shaderName.ReplaceExtension(""); // This will remove the ".shader" extension
+
+                AZ::IO::Path outputAzslFilePath = request.m_tempDirPath;
+                outputAzslFilePath /= AZStd::string::format("%s_%s_%s.azsl", materialTypeName.c_str(), materialPipelineName.c_str(), shaderName.c_str());
+
+                if (AZ::Utils::WriteFile(generatedAzsl, outputAzslFilePath.c_str()).IsSuccess())
+                {
+                    AssetBuilderSDK::JobProduct product;
+                    product.m_outputFlags = AssetBuilderSDK::ProductOutputFlags::IntermediateAsset;
+                    product.m_dependenciesHandled = true;
+                    product.m_productFileName = outputAzslFilePath.String();
+                    product.m_productSubID = nextProductSubID++;
+                    response.m_outputProducts.push_back(AZStd::move(product));
+                }
+                else
+                {
+                    AZ_Error(MaterialTypeBuilderName, false, "Failed to write intermediate azsl file '%s'.", outputAzslFilePath.c_str());
+                    return;
+                }
+
+                // Intermediate shader file
+
+                AZStd::string azslFileReference = AZ::IO::Path{outputAzslFilePath.Filename()}.c_str();
+                AZStd::to_lower(azslFileReference.begin(), azslFileReference.end());
+                AzFramework::StringFunc::Replace(shaderFile.GetValue(), "INSERT_AZSL_HERE", azslFileReference.c_str());
+
+                AZ::IO::Path outputShaderFilePath = request.m_tempDirPath;
+                outputShaderFilePath /= AZStd::string::format("%s_%s_%s.shader", materialTypeName.c_str(), materialPipelineName.c_str(), shaderName.c_str());
+
+                if (AZ::Utils::WriteFile(shaderFile.GetValue(), outputShaderFilePath.c_str()).IsSuccess())
```

Comment:
Instead of reading in the text for the shader file, I recommend loading the shader source data structure from the file, adding the file reference, and saving the output shader source data file. Then you don't need INSERT_AZSL_HERE.

Reply:
I kind of like having "INSERT_AZSL_HERE" since the file is a template and not a complete valid shader asset, this makes it clear to readers what aspect of the file is a template and will be filled in later.

Reply:
I see. The same idea could be conveyed by moving the generated code block with the placeholder material functions to its own file and setting that as the default path instead using INSERT_AZSL_HERE. It might even simplify the builder. In addition to that, since the generated file will now be a “real file”, the user could just include it at the end of their custom AZSL to provide the defaults for whatever functions they didn't implement.

Also, when I initially read this, INSERT_AZSL_HERE seemed like it was expecting code to be inserted into .shader file. Sticking with this token, I recommend making it something more explicit to convey that it's a path.

Reply:
The thing the actual entry point .azsl file does not exist, that's generated by the asset builder. Both the template .azsli file and the material type's .azsli file are just snippets. The builder generates the .azsl that stitches them together, and that's what the .shader needs to reference.

I can change the macro name in my next PR, so I don't have to re-run Jenkins AR.